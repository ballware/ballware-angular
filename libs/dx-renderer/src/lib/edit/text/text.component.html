<dx-text-box *ngIf="layoutItem && layoutItem.options && layoutItem.options.dataMember"
  [label]="layoutItem.options.caption ?? ''"
  [name]="layoutItem.options.dataMember"
  [(value)]="value"
  [readOnly]="(readonly$ | async) ?? false"
  [showClearButton]="true"
  (onFocusIn)="onEntered()">
  <dx-validator>
    <dxi-validation-rule *ngIf="(required$ | async) ?? false" type="required"></dxi-validation-rule>
  </dx-validator>
</dx-text-box>
