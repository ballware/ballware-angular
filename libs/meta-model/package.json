{
  "name": "@ballware/meta-model",
  "version": "8.0.1",
  "type": "commonjs"
}
