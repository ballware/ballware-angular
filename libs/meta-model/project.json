{
  "name": "meta-model",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/meta-model/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": [
        "{options.outputPath}"
      ],
      "options": {
        "outputPath": "dist/libs/meta-model",
        "main": "libs/meta-model/src/index.ts",
        "tsConfig": "libs/meta-model/tsconfig.lib.json",
        "assets": [
          "libs/meta-model/*.md"
        ]
      }
    },
    "publish": {
      "executor": "nx:run-commands",
      "options": {
        "command": "node tools/scripts/publish.mjs meta-model {args.ver} {args.tag}"
      },
      "dependsOn": [
        "build"
      ]
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": [
        "{options.outputFile}"
      ],
      "options": {
        "lintFilePatterns": [
          "libs/meta-model/**/*.ts"
        ]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": [
        "{workspaceRoot}/coverage/{projectRoot}"
      ],
      "options": {
        "jestConfig": "libs/meta-model/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
