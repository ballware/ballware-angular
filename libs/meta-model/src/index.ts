export * from './lib/cruditem';
export * from './lib/document';
export * from './lib/entity';
export * from './lib/page';
export * from './lib/queryparams';
export * from './lib/rights';
export * from './lib/scriptutil';
export * from './lib/statistic';
export * from './lib/template';
export * from './lib/tenant';

