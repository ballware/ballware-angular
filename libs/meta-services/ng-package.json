{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/libs/meta-services",
  "lib": {
    "entryFile": "src/index.ts"
  }
}
