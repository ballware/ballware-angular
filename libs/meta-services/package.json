{
  "name": "@ballware/meta-services",
  "version": "8.0.6",
  "peerDependencies": {
    "@angular/common": "^14.2.0",
    "@angular/core": "^14.2.0"
  },
  "dependencies": {
    "tslib": "^2.3.0"
  }
}
