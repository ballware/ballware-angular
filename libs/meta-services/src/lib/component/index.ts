import { StoreModule } from '@ngrx/store';
import { componentReducer } from './component.reducer';
import { componentFeatureKey } from './component.state';

export * from './component.actions';
export * from './component.state';

export const ComponentFeatureModule = StoreModule.forFeature(componentFeatureKey, componentReducer);