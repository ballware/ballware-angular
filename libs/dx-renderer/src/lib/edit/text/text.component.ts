import { Component, Inject, Input, OnInit } from '@angular/core';
import { EditLayoutItem } from '@ballware/meta-model';
import { EDIT_SERVICE, EditService } from '@ballware/meta-services';
import { takeUntil } from 'rxjs';
import { WithDestroy } from '../../utils/withdestroy';
import { WithEditItemLifecycle } from '../../utils/withedititemlivecycle';
import { WithReadonly } from '../../utils/withreadonly';
import { WithRequired } from '../../utils/withrequired';
import { WithValidation } from '../../utils/withvalidation';
import { WithValue } from '../../utils/withvalue';
import { WithVisible } from '../../utils/withvisible';
import { CommonModule } from '@angular/common';
import { DxTextBoxModule, DxValidatorModule } from 'devextreme-angular';

@Component({
  selector: 'ballware-edit-text',
  templateUrl: './text.component.html',
  styleUrls: ['./text.component.scss'],
  imports: [CommonModule, DxTextBoxModule, DxValidatorModule],
  standalone: true
})
export class EditLayoutTextComponent extends WithVisible(WithRequired(WithValidation(WithReadonly(WithValue(WithEditItemLifecycle(WithDestroy()), () => ""))))) implements OnInit {

  @Input() initialLayoutItem?: EditLayoutItem;

  public layoutItem: EditLayoutItem|undefined;

  constructor(@Inject(EDIT_SERVICE) private editService: EditService) {
    super();
  }

  ngOnInit(): void {
    if (this.initialLayoutItem) {
      this.initLifecycle(this.initialLayoutItem, this.editService, this);

      this.preparedLayoutItem$
        .pipe(takeUntil(this.destroy$))
        .subscribe((layoutItem) => {
          if (layoutItem) {
            this.initValue(layoutItem, this.editService);
            this.initReadonly(layoutItem, this.editService);
            this.initValidation(layoutItem, this.editService);
            this.initRequired(layoutItem, this.editService);
            this.initVisible(layoutItem);

            if (layoutItem.type === 'mail') {
              this.validateEmail(true);
            }

            this.layoutItem = layoutItem;
          }
        });
    }
  }
}
