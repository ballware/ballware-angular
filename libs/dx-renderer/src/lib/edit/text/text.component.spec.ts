import { ComponentFixture, TestBed } from '@angular/core/testing';

import { EditLayoutTextComponent } from './text.component';
import { Provider } from '@angular/core';
import { EDIT_SERVICE } from '@ballware/meta-services';
import { EditLayoutItem } from '@ballware/meta-model';
import { mockedEditServiceContext } from '../../../test/editservice.spec';

describe('EditLayoutTextComponent', () => {
  let component: EditLayoutTextComponent;
  let fixture: ComponentFixture<EditLayoutTextComponent>;

  const mockedEditService = mockedEditServiceContext();
        
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [ EditLayoutTextComponent ],
      providers: [        
        {
          provide: EDIT_SERVICE,
          useFactory: () => mockedEditService.mock.object()
        } as Provider         
      ]
    })
    .compileComponents();
  });

  it('should create text', () => {
    fixture = TestBed.createComponent(EditLayoutTextComponent);

    const layoutItem = {
      type: 'text',
      options: {
          dataMember: 'mockedmember',
      }
    } as EditLayoutItem;
    
    component = fixture.componentInstance;   
    expect(component).toBeTruthy();

    component.initialLayoutItem = layoutItem;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should create mail', () => {
    fixture = TestBed.createComponent(EditLayoutTextComponent);

    const layoutItem = {
      type: 'mail',
      options: {
          dataMember: 'mockedmember',
      }
    } as EditLayoutItem;
    
    component = fixture.componentInstance;   
    expect(component).toBeTruthy();

    component.initialLayoutItem = layoutItem;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should apply options', () => {
    fixture = TestBed.createComponent(EditLayoutTextComponent);

    const layoutItem = {
      options: {
          dataMember: 'mockedmember',
          required: false,
          readonly: false,            
          visible: false
      }
    } as EditLayoutItem;
    
    component = fixture.componentInstance;   
    expect(component).toBeTruthy();

    component.initialLayoutItem = layoutItem;
    fixture.detectChanges();

    expect(component.getOption('value')).toBe("");
    expect(component.getOption('required')).toBe(false);
    expect(component.getOption('readonly')).toBe(false);
    expect(component.getOption('visible')).toBe(false);

    component.setOption('value', 'some text');
    expect(component.getOption('value')).toBe('some text');

    component.setOption('required', true);
    expect(component.getOption('required')).toBe(true);

    component.setOption('readonly', true);
    expect(component.getOption('readonly')).toBe(true);

    component.setOption('visible', true);
    expect(component.getOption('visible')).toBe(true);

    expect(() => component.getOption('undefined')).toThrowError('Unsupported option <undefined>');
    expect(() => component.setOption('undefined', 'any value')).toThrowError('Unsupported option <undefined>');
  });
});
