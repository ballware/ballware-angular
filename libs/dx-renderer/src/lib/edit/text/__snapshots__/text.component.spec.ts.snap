// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditLayoutTextComponent should create mail 1`] = `
<ballware-edit-text
  currentValue$={[Function BehaviorSubject]}
  destroy$={[Function Subject]}
  editService={[Function ]}
  editorEntered$={[Function Subject]}
  editorEvent$={[Function Subject]}
  emailValidation$={[Function BehaviorSubject]}
  globalReadonly="false"
  initialLayoutItem={[Function Object]}
  layoutItem={[Function Object]}
  notifyValueChange$={[Function Subject]}
  onEntered={[Function Function]}
  onEvent={[Function Function]}
  preparedLayoutItem$={[Function BehaviorSubject]}
  readonly$={[Function BehaviorSubject]}
  refreshValueTrigger$={[Function Subject]}
  required$={[Function BehaviorSubject]}
  requiredValidation$={[Function BehaviorSubject]}
  validationRules$={[Function Observable]}
  visible$={[Function BehaviorSubject]}
>
  <dx-text-box
    class="mt-2 dx-validator dx-visibility-change-handler dx-show-invalid-badge dx-textbox dx-texteditor dx-show-clear-button dx-editor-outlined dx-texteditor-empty dx-widget"
  >
    <div
      class="dx-texteditor-container"
    >
      <div
        class="dx-texteditor-input-container"
      >
        <input
          autocomplete="off"
          class="dx-texteditor-input"
          name="mockedmember"
          role="textbox"
          tabindex="0"
          type="text"
        />
        <div
          class="dx-placeholder"
          data-dx_placeholder=""
        />
      </div>
      <div
        class="dx-texteditor-buttons-container"
      >
        <span
          class="dx-clear-button-area"
        >
          <span
            class="dx-icon dx-icon-clear"
          />
        </span>
      </div>
    </div>
  </dx-text-box>
</ballware-edit-text>
`;

exports[`EditLayoutTextComponent should create text 1`] = `
<ballware-edit-text
  currentValue$={[Function BehaviorSubject]}
  destroy$={[Function Subject]}
  editService={[Function ]}
  editorEntered$={[Function Subject]}
  editorEvent$={[Function Subject]}
  emailValidation$={[Function BehaviorSubject]}
  globalReadonly="false"
  initialLayoutItem={[Function Object]}
  layoutItem={[Function Object]}
  notifyValueChange$={[Function Subject]}
  onEntered={[Function Function]}
  onEvent={[Function Function]}
  preparedLayoutItem$={[Function BehaviorSubject]}
  readonly$={[Function BehaviorSubject]}
  refreshValueTrigger$={[Function Subject]}
  required$={[Function BehaviorSubject]}
  requiredValidation$={[Function BehaviorSubject]}
  validationRules$={[Function Observable]}
  visible$={[Function BehaviorSubject]}
>
  <dx-text-box
    class="mt-2 dx-validator dx-visibility-change-handler dx-show-invalid-badge dx-textbox dx-texteditor dx-show-clear-button dx-editor-outlined dx-texteditor-empty dx-widget"
  >
    <div
      class="dx-texteditor-container"
    >
      <div
        class="dx-texteditor-input-container"
      >
        <input
          autocomplete="off"
          class="dx-texteditor-input"
          name="mockedmember"
          role="textbox"
          tabindex="0"
          type="text"
        />
        <div
          class="dx-placeholder"
          data-dx_placeholder=""
        />
      </div>
      <div
        class="dx-texteditor-buttons-container"
      >
        <span
          class="dx-clear-button-area"
        >
          <span
            class="dx-icon dx-icon-clear"
          />
        </span>
      </div>
    </div>
  </dx-text-box>
</ballware-edit-text>
`;
